apps/wwaw/build_wwaw.py

```diff
@@ -0,0 +1,604 @@
+
+        Args:
+            url (str): The URL of the web page.
+            html (str): The raw HTML content of the web page.
+            existing_names (set, optional): A set of agent names already used. Ensures the result is unique.
+
+        Returns:
+            str: A clean, unique agent name suitable for use as an identifier.
+        """
+        if existing_names is None:
+            existing_names = set()
+        title = _extract_title_from_html(html)
+
+        # If no title is found, fall back to using the URL path or netloc for the agent name
+        if not title:
+            parsed = urlparse(url)
+            title = parsed.path.strip("/") or parsed.netloc
+
+        base = sub(AGENT_NAME_SANITIZE_REGEX, " ", title).lower()
+        base = sub(AGENT_NAME_HYPHENATE_REGEX, "-", base).strip("-")
+
+        # Truncate cleanly to max allowed length
+        base = base.encode("ascii", errors="ignore").decode()
+        parts = base.split("-")
+
+        # If already short, return it as-is
+        if len("-".join(parts)) <= self.MAX_NAME_LEN:
+            base = "-".join(parts)
+        else:
+            # Use a greedy truncation strategy by alternating between adding elements
+            # from the start and end of the parts list, preserving meaning from both ends.
+            front, back = [], []
+            total_len = 2  # Accounts for the extra hyphen we'll add between front and back
+            i, j = 0, len(parts) - 1
+            turn = True  # Start by taking from the front
+
+            while i <= j:
+                candidate = parts[i] if turn else parts[j]
+                # Check if adding this part would exceed the maximum name length
+                if total_len + len(candidate) + 1 > self.MAX_NAME_LEN:
+                    break
+                if turn:
+                    front.append(candidate)
+                    i += 1
+                else:
+                    back.insert(0, candidate)  # Add to the beginning of the back list
+                    j -= 1
+                total_len += len(candidate) + 1
+                turn = not turn  # Alternate between front and back
+
+            base = "-".join(front + ["-"] + back)
+
+        # If the name already exists, append a hash suffix to ensure uniqueness
+        if base in existing_names:
+            hash_suffix = md5(url.encode()).hexdigest()[:6]
+            base = (base[:self.MAX_NAME_LEN - 7].rstrip("-") + "-" + hash_suffix)
+
+        if len(base) > self.MAX_NAME_LEN:
+            raise ValueError(f"Agent name too long: '{base}', len={len(base)}>{self.MAX_NAME_LEN}.")
+
+        return base
+
+    def _process_page(
+            self,
+            url: str,
+            parent_name: Optional[str],
+            resp,
+            visited: set,
+            existing_names: set,
+            agents: dict,
+            count: int,
+            to_visit: List[Tuple[str, Optional[str]]],
+            base_domain: str
+    ) -> int:
```

Comment:
This method could eventually use some more comments, but I think I get what as going on.
This is basically scrape().